node/forest_libp2p/src/service.rs

```diff
@@ -63,57 +83,76 @@ impl Libp2pService {
             swarm.subscribe(topic);
         }
 
-        Libp2pService { swarm }
+        let (pubsub_sender_in, pubsub_receiver_in) = channel(20);
+        let (pubsub_sender_out, pubsub_receiver_out) = channel(20);
+        Libp2pService {
+            swarm,
+            pubsub_receiver_in,
+            pubsub_sender_in,
+            pubsub_receiver_out,
+            pubsub_sender_out,
+            log,
+        }
     }
-}
-
-impl Stream for Libp2pService {
-    type Item = NetworkEvent;
-    type Error = ();
 
-    /// Continuously polls the Libp2p swarm to get events
-    fn poll(&mut self) -> Result<Async<Option<Self::Item>>, Self::Error> {
+    /// Starts the `Libp2pService` networking stack. This Future resolves when shutdown occurs.
+    pub async fn run(self) -> Result<(), ()> {
+        enum MergeEvent {
+            Swarm(MyBehaviourEvent),
+            Pubsub((Topic, Vec<u8>)),
+        };
+        let mut swarm_stream = self.swarm.map(MergeEvent::Swarm).fuse();
+        let mut pubsub_stream = self.pubsub_receiver_in.map(MergeEvent::Pubsub).fuse();
         loop {
-            match self.swarm.poll() {
-                Ok(Async::Ready(Some(event))) => match event {
+            select! {
+                swarm_event = swarm_stream.next() => match swarm_event {
+                    Some(MergeEvent::Swarm(event)) => match event {
                     MyBehaviourEvent::DiscoveredPeer(peer) => {
-                        libp2p::Swarm::dial(&mut self.swarm, peer);
+                        libp2p::Swarm::dial(&mut swarm_stream.get_mut().get_mut(), peer);
                     }
                     MyBehaviourEvent::ExpiredPeer(_) => {}
                     MyBehaviourEvent::GossipMessage {
                         source,
                         topics,
                         message,
                     } => {
-                        return Ok(Async::Ready(Option::from(NetworkEvent::PubsubMessage {
+                        info!(self.log, "Got a Gossip Message from {:?}", source);
+                        self.pubsub_sender_out.send(NetworkEvent::PubsubMessage {
                             source,
                             topics,
-                            message,
-                        })));
+                            message
+                        }).await;
                     }
+                    }
+                    Some(MergeEvent::Pubsub(_)) => error!(self.log, "Should not happen"),
```

Comment:
Can you log a more detailed message? Would it be worth passing through the merge event to log it?

Reply:
I changed it to unreachable

Reply:
But isn't that case technically reachable? Unreachable macro should only be used for code that is literally impossible to reach, this doesn't seem like something we want to panic on

Reply:
This is literally unreachable. If you look at l110, I map all the Stream items to MergeEvent::Swarm